File: examples/scripts/ts/activity_analysis.ts
/**
 * Model Health TypeScript SDK — post-capture analysis workflow.
 * Mirrors examples/python/activity_analysis.py.
 *
 * Usage:
 *   npx tsx activity_analysis.ts [<api_key>]
 */

import { ModelHealthClient, ActivityType } from '@modelhealth/modelhealth';
import type { AnalysisDataType } from '@modelhealth/modelhealth';
import {
  loadApiKey, INTERNAL_ACTIVITY_NAMES, ANALYSIS_DATA_EXT,
  pickOne, pickMulti, saveFile, pollAnalysis, sleep, closePrompts,
} from './_shared.js';

const ANALYSIS_TYPES: [string, string][] = [
  [ActivityType.CounterMovementJump, 'Counter Movement Jump'],
  [ActivityType.Gait,                'Overground Walking'],
  [ActivityType.TreadmillGait,       'Treadmill Walking'],
  [ActivityType.TreadmillRunning,    'Treadmill Running'],
  [ActivityType.OvergroundRunning,   'Overground Running'],
  [ActivityType.SitToStand,          'Sit-to-Stand Transfer'],
  [ActivityType.Squats,              'Squats'],
  [ActivityType.RangeOfMotion,       'Range of Motion'],
  [ActivityType.DropJump,            'Drop Vertical Jump'],
  [ActivityType.Hop,                 'Hop Test'],
  [ActivityType.ChangeOfDirection,   '5-0-5 Test'],
  [ActivityType.Cut,                 'Cutting Maneuver'],
  [ActivityType.Sprint,              'Sprint'],
  [ActivityType.LateralStepdown,     'Lateral Step Down'],
  [ActivityType.Lunge,               'Lunge'],
];

const RESULT_TYPES: [AnalysisDataType, string][] = [
  ['report',  'Report   (PDF) '],
  ['data',    'Data     (ZIP) '],
];

async function connect(apiKey: string): Promise<ModelHealthClient> {
  console.log('Connecting...');
  const client = new ModelHealthClient({ apiKey, autoInit: false });
  await client.init();
  return client;
}

async function pickSession(client: ModelHealthClient) {
  console.log('\nFetching sessions...');
  const sessions = await client.sessionList();
  if (!sessions.length) {
    console.error('No sessions found. Create a session using the Model Health mobile app first.');
    process.exit(1);
  }

  console.log(`\n${sessions.length} session(s):\n`);
  return pickOne(sessions, 'Select session', s => {
    const sn = s.sessionName || '(unnamed)';
    const sub = s.name || '(unnamed)';
    return `[session ID: ${s.id}]  session name: ${sn}  subject: ${sub}`;
  });
}

async function pickActivity(client: ModelHealthClient, session: Awaited<ReturnType<typeof pickSession>>) {
  const sn = session.sessionName || '(unnamed)';
  const sub = session.name || '(unnamed)';
  console.log(`\nFetching activities for session ID: ${session.id},  session name: ${sn}, subject: ${sub}...`);
  const allActivities = await client.activityList(session.id);
  const activities = allActivities.filter(a => !INTERNAL_ACTIVITY_NAMES.has(a.name ?? ''));
  if (!activities.length) { console.error('No activities found in this session.'); process.exit(1); }

  console.log(`\n${activities.length} activity/activities:\n`);
  return pickOne(
    activities,
    'Select activity',
    a => `${a.name ?? a.id}  [${a.status}]` + (a.activityType ? `  ${a.activityType}` : '')
  );
}

async function ensureReady(client: ModelHealthClient, activity: Awaited<ReturnType<typeof pickActivity>>) {
  const activityLabel = activity.name ?? activity.id;
  console.log(`\nChecking status of '${activityLabel}'...`);
  let status = await client.activityStatus(activity);

  if (status.type === 'uploading' || status.type === 'processing') {
    console.log('Waiting for processing to complete...');
    status = await pollActivity(client, activity);
  }

  if (status.type !== 'ready') {
    console.error(`Activity cannot be analysed (status: ${status.type}). Wait for uploads to finish and try again.`);
    process.exit(1);
  }
  console.log('Activity is ready.');
}

async function startAnalysis(
  client: ModelHealthClient,
  activity: Awaited<ReturnType<typeof pickActivity>>,
  session: Awaited<ReturnType<typeof pickSession>>
) {
  // Default to the activity's recorded type if available.
  const defaultAnalysis = ANALYSIS_TYPES.find(t => t[0] === activity.activityType);
  console.log('\nAnalysis type:\n');
  const [analysisType, analysisLabel] = await pickOne(
    ANALYSIS_TYPES, 'Select analysis type', t => t[1], defaultAnalysis
  );

  console.log(`\nStarting '${analysisLabel}' analysis...`);
  return client.startAnalysis(analysisType as any, activity, session);
}

async function waitForAnalysis(
  client: ModelHealthClient,
  task: Awaited<ReturnType<typeof startAnalysis>>,
  activity: Awaited<ReturnType<typeof pickActivity>>
) {
  console.log('Waiting for analysis to complete...');
  const resultStatus = await pollAnalysis(client, task);

  if (resultStatus.type !== 'completed') {
    console.error(`Analysis did not complete (status: ${resultStatus.type}).`);
    process.exit(1);
  }
  console.log('Analysis complete.');

  return client.fetchActivity(activity.id);
}

async function downloadResults(client: ModelHealthClient, activity: Awaited<ReturnType<typeof waitForAnalysis>>) {
  console.log('\nWhich results would you like to save?\n');
  const selected = await pickMulti(RESULT_TYPES, 'Select result types', r => r[1]);
  const dataTypes = selected.map(r => r[0]);

  const slug = (activity.name ?? activity.id).replace(/ /g, '_');

  console.log('\nDownloading...');
  const results = await client.analysisDataForActivity(activity, dataTypes);
  for (const r of results) {
    const ext = ANALYSIS_DATA_EXT[r.type] ?? 'bin';
    const p = saveFile(`${slug}_${r.type}.${ext}`, r.data);
    console.log(`  Saved ${p}`);
  }
}

async function main() {
  const args = process.argv.slice(2);
  const client = await connect(loadApiKey(args[0]));
  const session = await pickSession(client);
  const activity = await pickActivity(client, session);
  await ensureReady(client, activity);

  const task = await startAnalysis(client, activity, session);
  const freshActivity = await waitForAnalysis(client, task, activity);
  await downloadResults(client, freshActivity);

  console.log('\nDone.');
}

async function pollActivity(
  client: ModelHealthClient,
  activity: Parameters<typeof client.activityStatus>[0]
) {
  while (true) {
    const status = await client.activityStatus(activity);
    if (status.type === 'uploading') {
      process.stdout.write(`  Uploading (${status.uploaded}/${status.total} cameras)...  \r`);
    } else if (status.type === 'processing') {
      process.stdout.write('  Processing...                              \r');
    } else {
      process.stdout.write('\n');
      return status;
    }
    await sleep(10_000);
  }
}

main().catch(err => { console.error(`Error: ${err.message ?? err}`); process.exit(1); })
  .finally(closePrompts);

File: examples/scripts/python/activity_analysis.py
#!/usr/bin/env python3
"""Model Health Python SDK — interactive CLI demo.

Walks through the post-capture workflow:
  1. Select a session
  2. Select an activity from that session
  3. Wait for processing to complete (if needed)
  4. Choose an analysis type and run it
  5. Choose which result files to save (report PDF, data ZIP)

Usage:
    activity_analysis.py [<api_key>]
"""

import sys
import time

from docopt import docopt

from modelhealth import (
    ModelHealthService,
    ModelHealthError,
    ActivityStatus,
    ActivityStatusUploading,
    AnalysisStatus,
    ActivityType,
    AnalysisDataType,
)
from _prompts import pick_one, pick_multi
from _utils import save_file, ANALYSIS_DATA_EXT, load_api_key, poll_analysis

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

ANALYSIS_TYPES = [
    (ActivityType.counter_movement_jump, "Counter Movement Jump"),
    (ActivityType.gait,                  "Overground Walking"),
    (ActivityType.treadmill_gait,        "Treadmill Walking"),
    (ActivityType.treadmill_running,     "Treadmill Running"),
    (ActivityType.overground_running,    "Overground Running"),
    (ActivityType.sit_to_stand,          "Sit-to-Stand Transfer"),
    (ActivityType.squats,                "Squats"),
    (ActivityType.range_of_motion,       "Range of Motion"),
    (ActivityType.drop_jump,             "Drop Vertical Jump"),
    (ActivityType.hop,                   "Hop Test"),
    (ActivityType.change_of_direction,   "5-0-5 Test"),
    (ActivityType.cut,                   "Cutting Maneuver"),
    (ActivityType.sprint,                "Sprint"),
    (ActivityType.lateral_stepdown,      "Lateral Step Down"),
    (ActivityType.lunge,                 "Lunge"),
]

RESULT_TYPES = [
    (AnalysisDataType.report,  "Report   (PDF) "),
    (AnalysisDataType.data,    "Data     (ZIP) "),
]


# Activities created by the mobile app for internal use — exclude from lists.
_INTERNAL_ACTIVITY_NAMES = {"calibration", "neutral"}

# ---------------------------------------------------------------------------
# Polling helpers
# ---------------------------------------------------------------------------

def _poll_processing(service, activity, interval=10):
    """Block until the activity finishes uploading and processing.

    Returns the final status (ActivityStatus.ready or ActivityStatus.failed).
    """
    while True:
        status = service.activity_status(activity)
        if isinstance(status, ActivityStatusUploading):
            print(f"  Uploading ({status.uploaded}/{status.total} cameras)...  ", end="\r", flush=True)
        elif status == ActivityStatus.processing:
            print("  Processing...                              ", end="\r", flush=True)
        else:
            print()  # clear the \r line
            return status
        time.sleep(interval)


# ---------------------------------------------------------------------------
# Setup
# ---------------------------------------------------------------------------

def _connect(api_key):
    print("Connecting...")
    try:
        return ModelHealthService(api_key)
    except ModelHealthError as exc:
        sys.exit(f"Failed to initialise: {exc}")


# ---------------------------------------------------------------------------
# Session / activity selection
# ---------------------------------------------------------------------------

def _pick_session(service):
    print("\nFetching sessions...")
    sessions = service.session_list()
    if not sessions:
        sys.exit(
            "No sessions found. Create a session using the Model Health mobile app first."
        )

    print(f"\n{len(sessions)} session(s):\n")
    return pick_one(
        sessions,
        "Select session",
        lambda s: f"[session ID: {s.id}]  session name: {s.session_name or '(unnamed)'}  subject: {s.name or '(unnamed)'}",
    )


def _pick_activity(service, session):
    print(f"\nFetching activities for session ID: {session.id},  session name: {session.session_name or '(unnamed)'}, subject: {session.name or '(unnamed)'}...")
    all_activities = service.activity_list(session)
    activities = [
        a for a in all_activities
        if a.name not in _INTERNAL_ACTIVITY_NAMES
    ]
    if not activities:
        sys.exit("No activities found in this session.")

    print(f"\n{len(activities)} activity/activities:\n")
    return pick_one(
        activities,
        "Select activity",
        lambda a: f"{a.name or a.id}  [{a.status}]" + (f"  {a.activity_type}" if a.activity_type else ""),
    )


# ---------------------------------------------------------------------------
# Readiness
# ---------------------------------------------------------------------------

def _ensure_ready(service, activity):
    activity_label = activity.name or activity.id
    print(f"\nChecking status of '{activity_label}'...")
    status = service.activity_status(activity)

    if isinstance(status, ActivityStatusUploading) or status == ActivityStatus.processing:
        print("Waiting for processing to complete...")
        status = _poll_processing(service, activity)

    if status != ActivityStatus.ready:
        sys.exit(
            f"Activity cannot be analysed (status: {status}). "
            "Wait for uploads to finish and try again."
        )
    print("Activity is ready.")


# ---------------------------------------------------------------------------
# Analysis
# ---------------------------------------------------------------------------

def _start_analysis(service, activity, session):
    # Default to the activity's recorded type if available.
    default_analysis = next(
        (t for t in ANALYSIS_TYPES if t[0] == activity.activity_type),
        None,
    )
    print("\nAnalysis type:\n")
    analysis_value, analysis_label = pick_one(
        ANALYSIS_TYPES,
        "Select analysis type",
        lambda t: t[1],
        default=default_analysis,
    )

    print(f"\nStarting '{analysis_label}' analysis...")
    try:
        return service.start_analysis(analysis_value, activity, session)
    except ModelHealthError as exc:
        sys.exit(f"Failed to start analysis: {exc}")


def _wait_for_analysis(service, task, activity):
    print("Waiting for analysis to complete...")
    result_status = poll_analysis(service, task)

    if result_status != AnalysisStatus.completed:
        sys.exit(f"Analysis did not complete (status: {result_status}).")
    print("Analysis complete.")

    # Re-fetch the activity so its results field contains the analysis URLs.
    return service.fetch_activity(activity.id)


# ---------------------------------------------------------------------------
# Results
# ---------------------------------------------------------------------------

def _download_results(service, activity):
    print("\nWhich results would you like to save?\n")
    selected = pick_multi(
        RESULT_TYPES,
        "Select result types",
        lambda r: r[1],
    )
    data_types = [r[0] for r in selected]

    slug = (activity.name or activity.id).replace(" ", "_")

    print("\nDownloading...")
    results = service.analysis_data_for_activity(activity, data_types)
    for r in results:
        ext = ANALYSIS_DATA_EXT.get(r.type, "bin")
        path = save_file(f"{slug}_{r.type}.{ext}", r.data)
        print(f"  Saved {path}")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main(api_key):
    service = _connect(api_key)
    session = _pick_session(service)
    activity = _pick_activity(service, session)
    _ensure_ready(service, activity)

    task = _start_analysis(service, activity, session)
    activity = _wait_for_analysis(service, task, activity)
    _download_results(service, activity)

    print("\nDone.")


if __name__ == "__main__":
    args = docopt(__doc__)
    main(load_api_key(args["<api_key>"]))
